### PR TITLE
Reduce memory usage in field-caps responses

### PR DESCRIPTION
**Pull request by dnhatn**
We have reduced the memory usage of field-caps requests targeting many indices in 8.2+ (see https://github.com/elastic/elasticsearch/pull/83494). Unfortunately, we still receive OOM reports in 7.17. I think we should push some contained improvements to reduce the memory usage for those requests in 7.17. I have looked into several options. This PR can reduce the memory usage of field-caps responses by replace HashMap with ArrayList for the field responses to eliminate duplicated string names and internal nodes of Map. I will look into other options, such as deduplicating field-name strings or field-responses.

### CODE REVIEW

**Comment by javanna** [comment]
Heya @nhat, this change makes sense to me, good catch! Do you have numbers on the improvements when it comes to memory usage? Did you assess whether the change also has some effect on response time of field_caps?

**Comment by dnhatn** [comment]
@javanna Thanks for looking. This PR doesn't reduce the latency significantly as we still send a full map over the wire. The current version hits OOM with 10 concurrent field-caps requests, but this change doesn't.

**Review by javanna — APPROVED** [review]
Asked a question around testing, LGTM otherwise. The reason why I asked about numbers is that it would be nice to understand what to expect with this changes. Is it a 1%, 10% or 50% memory usage reduction approximately?

File: server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesIndexResponse.java [hunk 1]
```diff
@@ -72,7 +92,7 @@ Version getOriginVersion() {
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         out.writeString(indexName);
-        out.writeMap(responseMap, StreamOutput::writeString, (valueOut, fc) -> fc.writeTo(valueOut));
+        writeFields(out, fields);
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesIndexResponse.java:95]: We have field-caps BWC tests and unit tests for node-level and cross-clusters responses which should cover FieldCapabilitiesIndexResponse. Anyway, I added a test for this class in https://github.com/elastic/elasticsearch/pull/88042/commits/b3f91be0b43a37df4ba3feba2194393b5b578585.

**Comment by dnhatn** [comment]
@elasticmachine test this please

**Comment by dnhatn** [comment]
> Is it a 1%, 10% or 50% memory usage reduction approximately?

I expect this PR reduces the memory usage by around 40%.

**Comment by dnhatn** [comment]
@javanna Thanks for review.

**Pull request merged**